src/pokeathlon/pokeathlon_course.c

```diff
@@ -87,3 +101,152 @@ BOOL PokeathlonCourse_Init(OverlayManager *manager, int *state) {
 
     return TRUE;
 }
+
+BOOL PokeathlonCourse_Main(OverlayManager *manager, int *state) {
+    PokeathlonCourseData *data;
+    PokeathlonCourseState *courseState;
+
+    data = OverlayManager_GetData(manager);
+    courseState = &data->courseState;
+
+    // Frame counter system: increments frameTimer each frame, and frameCounter every 0x708 frames (up to max 0xEA5F)
+    if (data->counterEnabled != 0) {
+        data->frameTimer++;
+        if (data->frameTimer >= 0x708) {
+            if (data->frameCounter < 0xEA5F) {
+                data->frameCounter++;
+            }
+            data->frameTimer = 0;
+        }
+    }
+
+    // Main state machine
+    switch (courseState->mainState) {
+    case 0: // Idle state - waiting for transitions
```

Comment:
normally for state machines like this we like to use enums for the switch cases
just to provide a bit of context as to the case